File: tools/feles-build.rb
#!/usr/bin/env ruby
# SPDX-License-Identifier: MPL-2.0
#
#   deno task feles-build <command>   (= ruby tools/feles-build.rb <command>)
#
# 建てる段を順に呼ぶだけ。中身は tools/lib/ に。

# mac の /usr/bin/ruby は 2.6。素っ気ない SyntaxError になる前に、どうすればいいか言っておく。
# (3.1 は endless method の書き方から。CI の ubuntu-latest は 3.2 なのでそのまま通る)
abort "ruby #{RUBY_VERSION} は古い(3.1 以上が要る)。`mise install` で mise.toml の版が入る" if RUBY_VERSION < "3.1"

require_relative "lib/browser_launcher"
require_relative "lib/builder"
require_relative "lib/dev_browser"
require_relative "lib/dev_env_manager"
require_relative "lib/dev_server"
require_relative "lib/initializer"
require_relative "lib/injector"
require_relative "lib/omni"
require_relative "lib/patcher"
require_relative "lib/symlinker"
require_relative "lib/update"
require_relative "lib/utils"
require_relative "lib/xhtml"

module FelesBuild
  LOGGER = Utils::Logger.new("feles-build")

  USAGE = <<~TXT
    Usage: deno task feles-build <command> [options]

    Commands:
      dev        Run the development workflow
      stage      Build production assets and run browser in dev mode
      build      Run the production build workflow (--phase before-mach|after-mach)
      stop       Close the dev browser started from this checkout
      misc       Misc commands (e.g. 'misc patch --action apply', 'misc writeVersion')
  TXT

  def self.build_id
    id = Update.build_id
    LOGGER.info "Generated build id: #{id}"
    id
  end

  # dev と stage で同じ道。omni.ja は、ほどいてから触って、最後に結ぶ。
  def self.prepare(mode)
    Initializer.run
    Omni.open
    Patcher.run("apply")
    Symlinker.run
    Builder.run(mode, build_id)
    Injector.run(mode)
    Xhtml.inject(Omni.root, dev: true)
    Omni.seal
    Initializer.resign_mac_app # 封を結び直すのは、omni.ja を最後に触ったあと
    DevEnvManager.setup
  end

  # 建てて、開いて、閉じたら一緒に片づける(vite だけ残さない)。
  def self.serve_and_launch
    %w[INT TERM HUP].each do |signal|
      Signal.trap(signal) do
        LOGGER.info "Shutting down (SIG#{signal})..."
        DevServer.shutdown
        exit 130
      end
    end

    DevServer.run
    LOGGER.success "Dev servers are ready."
    BrowserLauncher.run
  ensure
    LOGGER.info "Browser closed; stopping dev servers."
    DevServer.shutdown
  end

  def self.run_build(phase)
    case phase
    when "before-mach"
      Symlinker.run
      Builder.run("production", build_id)
    when "after-mach"
      Xhtml.inject(Defines::PROD_BIN_DIR, dev: false)
    when nil
      abort "Error: --phase is required for the build command."
    else
      abort "Unknown phase: #{phase}"
    end
  end

  def self.run_misc(argv)
    case argv.shift
    when "patch"
      action = flag(argv, "--action") || "apply"
      Omni.open # patch は omni の中にも当たるので、ほどいてから
      Patcher.run(action)
      Omni.seal if action == "apply"
    when "writeVersion" then Update.write_version("static/gecko")
    else abort USAGE
    end
  end

  def self.flag(argv, name)
    index = argv.index(name)
    index ? argv[index + 1] : nil
  end

  def self.main(argv)
    command = argv.shift
    return puts USAGE if argv.include?("--help") || argv.include?("-h")

    case command
    when "dev"
      LOGGER.info "Starting development environment..."
      prepare("dev")
      serve_and_launch
    when "stage"
      LOGGER.info "Starting staged production build (production assets) with browser in dev mode..."
      prepare("stage")
      serve_and_launch
    when "build" then run_build(flag(argv, "--phase"))
    when "stop" then DevBrowser.stop
    when "misc" then run_misc(argv)
    when "--help", "-h", nil then puts USAGE
    else
      LOGGER.error "Unknown command: #{command}"
      abort USAGE
    end
  end
end

begin
  FelesBuild.main(ARGV)
rescue => e
  FelesBuild::LOGGER.error "Unhandled error: #{e.message}"
  warn e.backtrace.first(5).join("\n") if ENV["DEBUG"]
  exit 1
end

File: tools/feles-build.ts
// SPDX-License-Identifier: MPL-2.0

import * as Initializer from "./src/initializer.ts";
import * as Patcher from "./src/patcher.ts";
import * as Builder from "./src/builder.ts";
import * as DevServer from "./src/dev_server.ts";
import * as Injector from "./src/injector.ts";
import * as BrowserLauncher from "./src/browser_launcher.ts";
import { Logger, runRuby, runRubyCapture } from "./src/utils.ts";

const logger = new Logger("feles-build");

/** build.rb uuid: Ruby の SecureRandom.uuid_v7 を一つもらう */
function generateBuildid2(): string {
  const buildid2 = runRubyCapture("uuid");
  logger.info(`Generated build id: ${buildid2}`);
  return buildid2;
}

/**
 * Simple writable-like object to capture ready signal from DevServer.run
 */
class ReadyPipe {
  private listeners: Array<(chunk: string) => void> = [];
  on(event: "data", cb: (chunk: string) => void) {
    if (event === "data") this.listeners.push(cb);
  }
  write(chunk: Uint8Array | string) {
    const s =
      typeof chunk === "string" ? chunk : new TextDecoder().decode(chunk);
    for (const cb of this.listeners) cb(s);
  }
  end() {
    // no-op
  }
}

/**
 * Common setup for dev and stage modes: graceful shutdown, dev server, and browser launch
 */
function setupDevServerAndBrowser(): void {
  // Graceful shutdown(Ctrl-C も、kill も、端末が閉じたときも。vite を残さない)
  const stop = (sig: string, code: number) => () => {
    logger.info(`Shutting down (${sig})...`);
    DevServer.shutdown();
    Deno.exit(code);
  };
  Deno.addSignalListener("SIGINT", stop("SIGINT", 130));
  Deno.addSignalListener("SIGTERM", stop("SIGTERM", 143));
  Deno.addSignalListener("SIGHUP", stop("SIGHUP", 129));

  const pipe = new ReadyPipe();

  pipe.on("data", (chunk: string) => {
    const s = chunk.toString().trim();
    if (
      s === (DevServer as any).DEV_SERVER?.ready_string ||
      s.includes("nora-")
    ) {
      logger.success("Dev servers are ready.");
      // Launch browser。閉じたら vite も止めて、この process も終わる(vite だけ残さない)
      BrowserLauncher.run()
        .catch((e: any) => {
          logger.error(`Browser launcher failed: ${e?.message ?? e}`);
        })
        .finally(() => {
          logger.info("Browser closed; stopping dev servers.");
          DevServer.shutdown();
          Deno.exit(0);
        });
    }
  });

  // Start dev server (it will write to the writer and end it)
  // DevServer.run expects a writable-like object.
  DevServer.run(pipe as any).catch((e: any) => {
    logger.error(`Dev server failed: ${e?.message ?? e}`);
    Deno.exit(1);
  });
}

async function runDev(): Promise<void> {
  logger.info("Starting development environment...");

  // Initial setup
  await Initializer.run();
  await Patcher.run("apply");
  runRuby("symlink");

  const buildid2 = generateBuildid2();
  await Builder.run("dev", buildid2);
  await Injector.run("dev");
  await Injector.injectXhtmlFromTs(true);
  // Re-seal the bundle after the last omni.ja modification (stock Firefox / darwin).
  Initializer.resignMacApp();
  runRuby("dev-env");

  setupDevServerAndBrowser();
  // Keep process alive until SIGINT or process termination from BrowserLauncher path
}

async function runStage(): Promise<void> {
  logger.info(
    "Starting staged production build (production assets) with browser in dev mode...",
  );

  // Initial setup
  await Initializer.run();
  await Patcher.run("apply");
  runRuby("symlink");
  const buildid2 = generateBuildid2();
  await Builder.run("stage", buildid2);
  await Injector.run("stage");
  await Injector.injectXhtmlFromTs(true);
  Initializer.resignMacApp();
  runRuby("dev-env");

  setupDevServerAndBrowser();
  // Keep process alive until SIGINT or browser termination like runDev
}

async function runBuild(phase?: string): Promise<void> {
  const optionsPhase = phase ?? null;
  if (!optionsPhase) {
    console.error("Error: --phase is required for the build command.");
    Deno.exit(1);
  }

  if (optionsPhase === "before-mach") {
    runRuby("symlink");
    // Build production assets
    const buildid2 = generateBuildid2();
    await Builder.run("production", buildid2);
  } else if (optionsPhase === "after-mach") {
    await Injector.injectXhtmlFromTs(false, true);
  } else {
    console.error(`Unknown phase: ${optionsPhase}`);
    Deno.exit(1);
  }
}

async function runPatch(action = "apply"): Promise<void> {
  await Patcher.run(action);
}

function printHelp(): void {
  console.log("Usage: deno task feles-build <command> [options]\n");
  console.log("Commands:");
  console.log("  dev        Run the development workflow");
  console.log(
    "  stage      Build production assets and run browser in dev mode",
  );
  console.log("  build      Run the production build workflow (--phase)");
  console.log("  misc       Misc commands (e.g. 'misc patch')");
  console.log("");
  console.log("Run 'feles-build <command> --help' for command-specific help.");
}

async function main(): Promise<void> {
  const argv = Deno.args.slice();
  const command = argv.shift();

  switch (command) {
    case "dev": {
      // simple help support
      if (argv.includes("--help") || argv.includes("-h")) {
        console.log("Usage: feles-build dev");
        return;
      }
      await runDev();
      break;
    }
    case "stage": {
      if (argv.includes("--help") || argv.includes("-h")) {
        console.log("Usage: feles-build stage");
        return;
      }
      await runStage();
      break;
    }
    case "build": {
      if (argv.includes("--help") || argv.includes("-h")) {
        console.log(
          "Usage: feles-build build --phase <before-mach|after-mach>",
        );
        return;
      }
      const idx = argv.indexOf("--phase");
      const phase = idx >= 0 ? argv[idx + 1] : undefined;
      await runBuild(phase);
      break;
    }
    case "misc": {
      if (argv.includes("--help") || argv.includes("-h")) {
        console.log(
          "Usage: feles-build misc patch --action <apply|create|init>",
        );
        return;
      }
      const sub = argv[0];
      if (sub === "patch") {
        const idx = argv.indexOf("--action");
        const action = idx >= 0 ? argv[idx + 1] : "apply";
        runPatch(action);
      } else if (sub === "writeVersion") {
        runRuby("write-version", "static/gecko");
      } else {
        logger.error(`Unknown misc command: ${sub}`);
        printHelp();
        Deno.exit(1);
      }
      break;
    }
    case "--help":
    case "-h":
    case undefined:
      printHelp();
      break;
    default:
      logger.error(`Unknown command: ${command}`);
      printHelp();
      Deno.exit(1);
  }
}

main().catch((e: any) => {
  logger.error(`Unhandled error: ${e?.message ?? e}`);
  Deno.exit(1);
});
